### PR TITLE
getting online nomads nemsio data

### PR DESCRIPTION
**Pull request by panll**
getting online data from nomad

modified file:
scripts/exregional_get_extrn_files.sh
new file:
ush/get_files.sh

nemsio files tested on hera
grib2 files needs updated chgres

### CODE REVIEW

File: scripts/exregional_get_extrn_files.sh [hunk 1]
```diff
@@ -595,6 +600,34 @@ In directory:    \"${scrfunc_dir}\"
 ========================================================================"
 
   fi
+elif [ "${DATA_SRC}" = "online" ]; then
+    print_info_msg "
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_files.sh]: For consistency with rest of code, can you do 2 spaces per indentation level?

File: scripts/exregional_get_extrn_files.sh [hunk 2]
```diff
@@ -595,6 +600,34 @@ In directory:    \"${scrfunc_dir}\"
...
+#
+#-----------------------------------------------------------------------
+#
+  prefix=${EXTRN_MDL_ARCVREL_DIR:+${EXTRN_MDL_ARCVREL_DIR}/}
+  EXTRN_MDL_FPS=( "${EXTRN_MDL_FNS[@]/#/$prefix}" )
+
+  EXTRN_MDL_FPS_str="( "$( printf "\"%s\" " "${EXTRN_MDL_FPS[@]}" )")"
+
+  echo "Getting online file EXTRN_MDL_FPS= ${EXTRN_MDL_FPS[@]}"
+    num_files_to_extract="${#EXTRN_MDL_FPS[@]}"
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_files.sh]: Again, please use 2 spaces instead of 4 when indenting to be consistent with the rest of the code (lines 623 to 629).

File: ush/get_files.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
```
Comment (gsketefian) [inline ush/get_files.sh]: Since this file just executes a single command (wget ...), can we just include that in exregional_get_extrn_files.sh instead of creating and calling this script?

File: scripts/exregional_make_ics.sh [hunk 4]
```diff
@@ -101,7 +101,7 @@ case "${CCPP_PHYS_SUITE}" in
   phys_suite="GFS"
   ;;
 
-"FV3_GSD_v0" | "FV3_GSD_SAR" | "FV3_GSD_SAR_v1" )
+"FV3_GSD_v0" | "FV3_GSD_SAR" | "FV3_GSD_SAR_v1" |"FV3_RRFS_v0" )
```
Comment (gsketefian) [inline scripts/exregional_make_ics.sh]: In order for the PRs to be more atomic (i..e contain only one issue instead of 2 or more unrelated issues), we should separate the changes that add the FV3_RRFS_v0 suite from the changes that are for getting files from online sources.

File: scripts/exregional_get_extrn_files.sh [hunk 5]
```diff
@@ -148,7 +148,24 @@ if [ "${num_files_found_on_disk}" -eq "${num_files_to_copy}" ]; then
   DATA_SRC="disk"
 else
   DATA_SRC="HPSS"
+  if [ "${MACHINE}" = "JET" ]; then
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_files.sh]: In a previous PR (#221) I added the capability to have different external file names depending on whether the source is disk or HPSS.  To keep the setting of machine-dependent variables localized as much as possible, we set all such variables in the file ush/get_extrn_mdl_dir_file_info.sh.  If this code is still needed (maybe #221 does what you need), please move to that script.

File: ush/get_files.sh (hunk 3 above)
Comment (panll) [reply ush/get_files.sh]: Yes, we can put it in the exregional_get_extrn_files.sh. Thanks! @gsketefian

File: scripts/exregional_get_extrn_files.sh (hunk 1 above)
Comment (panll) [reply scripts/exregional_get_extrn_files.sh]: Thanks, I'll change that.

File: scripts/exregional_get_extrn_files.sh (hunk 5 above)
Comment (panll) [reply scripts/exregional_get_extrn_files.sh]: I'll modify that.

File: ush/get_layout.sh [hunk 6]
```diff
@@ -0,0 +1,147 @@
+#!/bin/bash -f
```
Comment (gsketefian) [inline ush/get_layout.sh]: Can we separate the addition of this file into a new PR since it doesn't seem related to the other changes?  Then there can be more logical testing of each PR.

File: scripts/exregional_make_ics.sh (hunk 4 above)
Comment (panll) [reply scripts/exregional_make_ics.sh]: Yes, I can split that. Thanks!

**Comment by panll** [comment]
Files updated and conflicts are solved.

**Comment by panll** [comment]
get_layout.sh is removed. Tested on hera. @gsketefian

**Comment by gsketefian** [comment]
Hi Linlin,
Can you split these into two PRs, one for the changes needed in exregional_get_extrn_mdl_files.sh for getting data online and another for adding the RRFS_v0 suite?  Also, when you make the RRFS_v0 PR, can you add a WE2E test for that suite?  Thanks.

**Comment by panll** [comment]
Removed the files related to RRFS_v0 PR. Thanks! @gsketefian

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 7]
```diff
@@ -632,6 +637,33 @@ In directory:    \"${scrfunc_dir}\"
...
+#
+# Reset EXTRN_MDL_FPS to the full paths within the archive files of the
+# external model output files.
+#
+#-----------------------------------------------------------------------
+#
+  prefix=${extrn_mdl_arcvrel_dir:+${extrn_mdl_arcvrel_dir}/}
+  extrn_mdl_fps=( "${extrn_mdl_fns_on_disk[@]/#/$prefix}" )
+
+  extrn_mdl_fps_str="( "$( printf "\"%s\" " "${extrn_mdl_fps[@]}" )")"
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_mdl_files.sh:660]: The variable extrn_mdl_fps_str is not used.  To be consistent with the rest of the code, you can use this variable by replacing the echo statement with:

print_info_msg "
Getting external model files from nomads:
  extrn_mdl_fps = ${extrn_mdl_fps_str}"

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 8]
```diff
@@ -152,7 +152,12 @@ if [ "${num_files_found_on_disk}" -eq "${num_files_to_copy}" ]; then
   data_src="disk"
 else
   data_src="HPSS"
+  arcv_fp="${extrn_mdl_arcv_fps[0]}"
+  hsi ls $arcv_fp || data_src="online"
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_mdl_files.sh:160]: Can you add a comment to explain what you're doing here? I.e. checking if the first archive file exists on HPSS, and if not, setting data_src to online to try to get the files from nomads.

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 9]
```diff
@@ -152,7 +152,12 @@ if [ "${num_files_found_on_disk}" -eq "${num_files_to_copy}" ]; then
...
+  echo "ARCV_FP= $arcv_fp"
+  echo "DATA_SRC = $data_src"
 fi
+#data_src="online"
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_mdl_files.sh]: Remove this line?

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 10]
```diff
@@ -632,6 +637,33 @@ In directory:    \"${scrfunc_dir}\"
...
+elif [ "${data_src}" = "online" ]; then
+    print_info_msg "
+========================================================================
+getting data from online data sources
+========================================================================"
+
+#
+#-----------------------------------------------------------------------
+#
+# Reset EXTRN_MDL_FPS to the full paths within the archive files of the
```
Comment (gsketefian) [inline scripts/exregional_get_extrn_mdl_files.sh]: Change "Reset EXTRN_MDL_FPS" to "Set extrn_mdl_fps".

**Comment by gsketefian** [comment]
@panll I made a few minor comments.  Hopefully this will be the last set of mods for this PR.  Thanks.

**Comment by panll** [comment]
I'll modify it. Thanks! @gsketefian

**Comment by panll** [comment]
exregional_get_extrn_mdl_files.sh is modified according to your comments. @gsketefian

**Comment by gsketefian** [comment]
@panll Can you change the title of the PR to:
1) Remove mention of grib2 data (since it doesn't work with grib2 yet, right?); 
2) To be clear, mention "nomads" specifically instead of just "online".
3) Say that it has only been tested on hera.  

If you want it tested on other machines, please ask others.  Otherwise, I'll approve once you change the PR title.  Thx.

**Comment by mkavulich** [comment]
I'm testing the new capability on Cheyenne right now...the wget process is taking a long time (for a 12 hour forecast), I had to extend the wallclock time beyond 45 minutes and it is now re-running.

**Comment by panll** [comment]
Updated according to the comments. Thanks! @gsketefian  The nemsio file is around 15GB, and it needs a long time if the network is slow. Thanks! @mkavulich

**Comment by JeffBeck-NOAA** [comment]
Is there a status update on commits and testing for this PR?  Did testing succeed on multiple platforms?  Thanks!

**Comment by mkavulich** [comment]
I have still not successfully gotten this to work for cheyenne. There is something strange going on: the task hangs and will use the entire wallclock time, even if the file it is attempting to fetch from nomads does not exist. I also was able to manually download the files in just 20 minutes on Cheyenne, so I do not believe the download time is causing this issue. I'm still investigating; I turned everything to "verbose" on the latest run to hopefully get more information out.

**Comment by JulieSchramm** [comment]
There are issues with the NOMADS web site:  https://nomads.ncdc.noaa.gov.
My browser usually says "Can't find the server."

**Comment by JeffBeck-NOAA** [comment]
@mkavulich @JulieSchramm Would it be useful if @panll adds a "check" to exit immediately if the file isn't found, or if the connections errors out, and provides the "Can't find the server" message?

- Jeff

**Comment by mkavulich** [comment]
I've gotten a little closer to figuring out the actual problem. This is the contents of one of my log.wget.txt files when I disabled the "-nv" (not verbose) option:

```--2020-06-22 15:54:41--  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Resolving nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)... 140.172.138.18
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 16:08:56--  (try: 2)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 16:23:13--  (try: 3)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 16:37:32--  (try: 4)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 16:51:51--  (try: 5)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 17:06:13--  (try: 6)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 17:20:34--  (try: 7)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 17:34:55--  (try: 8)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443... failed: Connection timed out.
Retrying.

--2020-06-22 17:49:18--  (try: 9)  https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod/gfs.20200621/12/gfs.t12z.atmanl.nemsio
Connecting to nomads.ncep.noaa.gov (nomads.ncep.noaa.gov)|140.172.138.18|:443...
```

It appears as if the connection is simply failing for some reason. This only happens in the submitted job: if I do the same commands on the command line they are successful (if slow). Could it be there is some kind of block keeping us from using wget in the Cheyenne queue? Or perhaps the nomads server is blocking these connections on their end?

I'll open a cisl help ticket to see if they have any information.

**Comment by JeffBeck-NOAA** [comment]
@mkavulich That definitely sounds like there may be restrictions on the compute nodes (wget may just be disabled when running on the batch system).  We'll see what CISL says.

**Comment by panll** [comment]
The ncdc website issue has been reported on UFS forum several weeks ago:
https://forums.ufscommunity.org/threads/website-downloading-raw-ics-unavailable
The current link in wget is  only for very short time data. 
We can update that when more data are available.
@JulieSchramm @JeffBeck-NOAA

**Comment by mkavulich** [comment]
I got a response from CISL: the compute nodes are not connected to the internet, which is what is causing this issue. So we either need to submit these jobs to casper nodes (which I believe is not possible with the current rocoto setup) or download the data separately using the script outside of the workflow.

**Comment by RatkoVasic-NOAA** [comment]
> I got a response from CISL: the compute nodes are not connected to the internet, which is what is causing this issue. So we either need to submit these jobs to casper nodes (which I believe is not possible with the current rocoto setup) or download the data separately using the script outside of the workflow.

**Comment by RatkoVasic-NOAA** [comment]
> > I got a response from CISL: the compute nodes are not connected to the internet, which is what is causing this issue. So we either need to submit these jobs to casper nodes (which I believe is not possible with the current rocoto setup) or download the data separately using the script outside of the workflow.

**Comment by RatkoVasic-NOAA** [comment]
> > > I got a response from CISL: the compute nodes are not connected to the internet, which is what is causing this issue. So we either need to submit these jobs to casper nodes (which I believe is not possible with the current rocoto setup) or download the data separately using the script outside of the workflow.

Sorry, pressing the wrong button :-(
